libfoundation/src/foundation-number.cpp

```diff
@@ -216,41 +382,1247 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+typedef int64_t bigint_t;
+
+struct __MCNumberOperationAdd
+{
+    static inline double double_by_double(double x, double y)
+    {
+        return x + y;
+    }
+ 
+    static inline bigint_t bigint_by_bigint(bigint_t x, bigint_t y)
+    {
+        return x + y;
+    }
+    
+    static inline bool signed_by_signed(integer_t x, integer_t y, integer_t& r_value)
+    {
+        return __checked_signed_add(x, y, &r_value);
+    }
+    
+    static inline bool unsigned_by_unsigned(uinteger_t x, uinteger_t y, uinteger_t& r_value)
+    {
+        return __checked_unsigned_add(x, y, &r_value);
+    }
+    
+    static inline bool signed_by_unsigned(integer_t x, uinteger_t y, integer_t& r_value)
+    {
+        // As we are guaranteed that y will be in the range (INT_MAX, UINT_MAX]
+        // if x is positive, we cannot produce an integer_t.
+        if (x > 0)
+            return false;
+        
+        // if x <= 0 then
+        //   x + y === -(-x) + y === y - (-x)
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(y, (uinteger_t)-x, &z))
+            return false;
+        
+        if (z > INTEGER_MAX)
+            return false;
+        
+        r_value = (integer_t)z;
+        
+        return true;
+    }
+    
+    static inline bool unsigned_by_signed(uinteger_t x, integer_t y, integer_t& r_value)
+    {
+        // As we are guaranteed that x will be in the range (INT_MAX, UINT_MAX]
+        // if y is positive, we cannot produce an integer_t.
+        if (y > 0)
+            return false;
+        
+        // if y <= 0 then
+        //   x + y === x + -(-y) === x - (-y)
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(x, __negate_negative_signed(y), &z))
+            return false;
+        
+        if (z > INTEGER_MAX)
+            return false;
+        
+        r_value = (integer_t)z;
+        
+        return true;
+    }
+};
+
+struct __MCNumberOperationSubtract
+{
+    static inline double double_by_double(double x, double y)
+    {
+        return x - y;
+    }
+    
+    static inline bigint_t bigint_by_bigint(bigint_t x, bigint_t y)
+    {
+        return x - y;
+    }
+    
+    static inline bool signed_by_signed(integer_t x, integer_t y, integer_t& r_value)
+    {
+        return __checked_signed_subtract(x, y, &r_value);
+    }
+    
+    static inline bool unsigned_by_unsigned(uinteger_t x, uinteger_t y, uinteger_t& r_value)
+    {
+        return __checked_unsigned_subtract(x, y, &r_value);
+    }
+    
+    static inline bool signed_by_unsigned(integer_t x, uinteger_t y, integer_t& r_value)
+    {
+        // As we are guaranteed that y will be in the range (INT_MAX, UINT_MAX]
+        // if x is negative, we cannot produce an integer_t
+        if (x < 0)
+            return false;
+        
+        // x - y === -(y - x)
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(y, x, &z))
+            return false;
+        
+        if (z > -INTEGER_MIN)
```

Comment:
This unfortunately overflows. I think you need:

    if (z > -(int64_t)INTEGER_MIN)